### PR TITLE
[Relay][Quantize] Integrate data-aware calibration into quantization

### PR DESCRIPTION
**Pull request by vinx13**
This PR did some refactor work for the calibration part: integrate evaluation script for KL, the collect stats part has been moved into internal collect_stats. New config options `calibration_mode`, `weight_scale` have been added.
Removed `opt_level=3` for `prerequisite_optimize`, as it caused some accuracy issue when `FoldScaleAxis` is invoked before calibration.

Part of this PR is based on #3828.

@ZihengJiang @anijain2305 @tmoreau89

### CODE REVIEW

File: python/tvm/relay/quantize/_calibrate.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    ----------
+    graph: Function
+        The simulation graph after annotation.
+
+    Returns
+    -------
+    ret: Function
+        The profile graph which outputs a tuple of profile data.
+    """
+    logging.info("collecting statistics for calibration...")
```
Comment (anijain2305) [inline python/tvm/relay/quantize/_calibrate.py]: Nit - Blank line before here.

File: python/tvm/relay/quantize/quantize.py [hunk 2]
```diff
@@ -402,23 +304,21 @@ def prerequisite_optimize(graph, params=None):
...
+def quantize(mod, params=None, dataset=None):
     """ The quantization procedure. Before running the three main
     procedure of quantization, "annotate", "calibrate" and "realize"
     , we need to do "SimplifyInference", "FoldScaleAxis", "FoldConstant"
     first for optimizing.
 
     Parameters
     ---------
-    graph: Function
+    mod: Function
```
Comment (yzhliu) [inline python/tvm/relay/quantize/quantize.py]:
```suggestion
    mod: Module
```

File: python/tvm/relay/quantize/_calibrate.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+def collect_stats(mod, dataset):
+    """Given an annotated graph, create a profile graph to collect profile data from the
+    calibration dataset. This pass collects simulated_quantize op input into a tuple.
+    Simulated_quantize ops are rewritten to identity mode. The tuple is the output of the profile
+    graph.
+
+    Parameters
+    ----------
+    graph: Function
```
Comment (yzhliu) [inline python/tvm/relay/quantize/_calibrate.py]:
```suggestion
    mod: Module
```

File: python/tvm/relay/quantize/_calibrate.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        The simulation graph after annotation.
+
+    Returns
+    -------
+    ret: Function
```
Comment (yzhliu) [inline python/tvm/relay/quantize/_calibrate.py]: list of data?

**Comment by vinx13** [comment]
It's difficult to do unit tests. The refactor is covered by nightly tests. I plan to add more nightly tests for the new calibration mode later because there are some more work to do

**Comment by vinx13** [comment]
@yzhliu comments addressed

File: python/tvm/relay/quantize/kl_divergence.py [hunk 5]
```diff
@@ -54,6 +54,8 @@ def kl_divergence_scale(arr, quantized_dtype='int8', num_bins=8001, num_quantize
     http://on-demand.gputechconf.com/gtc/2017/presentation/s7310-8-bit-inference-with-tensorrt.pdf
     """
     assert isinstance(arr, np.ndarray)
+    assert stats is not None, "scipy need to be installed for \
```
Comment (tmoreau89) [inline python/tvm/relay/quantize/kl_divergence.py]: needs

File: python/tvm/relay/quantize/quantize.py [hunk 6]
```diff
@@ -143,9 +141,20 @@ def qconfig(**kwargs):
     nbit_dict: dict of QAnnotateKind -> int
         Number of bit for every kind of annotate field.
 
+    calibrate_mode: str
+        The calibration mode. 'global_scale' or 'kl'.
```
Comment (tmoreau89) [inline python/tvm/relay/quantize/quantize.py]: can we spell it out so it's more self explanatory (e.g. `kullback_leibler`)?

**Comment by tmoreau89** [comment]
@vinx13 I realize that this PR might have broken the E2E VTA tutorial flow (that includes quantization for VTA hardware pipeline). I'll need to investigate, but what puzzles me more is why that wasn't caught by the CI which should be building the sphinx galleries successfully, which should include the e2e VTA tutorial.

**Comment by tmoreau89** [comment]
The script in question is `vta/tutorials/frontend/deploy_vision_on_vta.py`

**Comment by tmoreau89** [comment]
Fix for the VTA bug introduced is in #4433.

**Pull request merged**